### ISSUE TITLE
Correct way to use Future and Barriers

### ISSUE THREAD

**Issue by ankuriitk** [ruby-concurrency/concurrent-ruby#610]
Hi,

I have a bunch of tasks (read ruby Procs) that I put in a queue. I then iterate over them and create futures for them. Then I have a barrier that waits for all these futures to finish. Code snippet is below. When I wrote a simple test for it, it seems to deadlock unless I use an immediate executor. I am not sure what I am doing wrong. Will appreciate if someone can spot check the code:

```ruby
 def do_work_with_futures
      futures = []
      futures_to_task_names = {}
      start_time = Time.now
      error = false
      info = {}
      until @workq.empty?
        p = @workq.pop(false)
        opts = { excecutor: GLOBAL_CUSTOM_EXECUTOR[@level] }
        ## If I use immediate executor then test passes but otherwise it deadlocks
        f = Concurrent::Future.execute(opts) { p.first.call }
        futures << f
        futures_to_task_names[f.object_id] = p.second
      end
      barrier = Concurrent.dataflow(*futures) do
        futures.each do |f|
          ## byebug at this line NEVER comes. 
          task_name = futures_to_task_names[f.object_id]
          puts "future #{task_name} state: #{f.state}, reason: #{f.reason}"
          # either the future didn't fulfill or it returned true. Both means error
          ferror = (f.state != :fulfilled) || f.value
          info[task_name] = { error: ferror, time: 0 }
          error ||= ferror
          if ferror
            error = true
            if f.reason.present?
              info[task_name][:error_message] = f.reason.message
              info[task_name][:stacktrace] = f.reason.backtrace.inspect
            end
          end
        end
      end

      barrier.wait_or_cancel(MAX_TIME_IN_SECONDS)
      puts "barrier value: #{barrier.value}"
```

**Comment by pitr-ch** [comment]
Could you simplify the code snippet just to essentials which do not work? Please include the executor configuration and how the queue is filled. It would help greatly if I could execute it locally.

**Comment by ankuriitk** [comment]
Hi Petr,

Apologies for the deafening silence on this. Here is a condensed and relevant info (including executor config). My question is if this is the correct way to use nested futures and barriers.

```
 module BackgroundThreads
  executors = []
  executors << Concurrent::ThreadPoolExecutor.new(
    max_threads: 20,
    min_threads: 4,
    max_queue: 2_147_483_647)

  executors << Concurrent::ThreadPoolExecutor.new(
    max_threads: 60,
    min_threads: 4,
    max_queue: 2_147_483_647)

  5.times do
    executors << Concurrent::ThreadPoolExecutor.new(
    max_threads: 10,
    min_threads: 0,
    max_queue: 2_147_483_647)
  end

  GLOBAL_ADWYZE_EXECUTOR = executors.freeze

class BGThreadPool

  def initialize( level = 0)
      @workq = Queue.new
      # Level is incremented for level of nesting of futures
      @level = level  
  end

 def addTask(proc, info)
      @workq.push [proc, info]
    end

 def do_work_with_futures
      futures = []
      until @workq.empty?
        p = @workq.pop(false)
        opts = { excecutor: GLOBAL_ADWYZE_EXECUTOR[@level] }
        f = Concurrent::Future.execute(opts) { p.first.call }
        futures << f
      end
      barrier = Concurrent.dataflow(*futures) do
        futures.each do |f|
          # do future specific processing to check if the future succeeded or not
        end
      end

      barrier.wait_or_cancel(MAX_TIME_IN_SECONDS)
    end
end

parent = BGThreadPool.new(0)
parent.addTask([proc1, "task1"])
parent.addTask([proc2, "task2"])

# See how proc itself is using its own instance of BGThreadPool and this is were the nesting of futures come
proc1 = Proc.new do
  child = BGThreadPool.new(1)
  child.addTask([anotherProc, "another task"])
  child.addTask([yetAnotherProc, "yet another task])
end

# proc2 = a similar proc with nested BGThreadPool
```

**Comment by pawandubey** [comment]
@pitr-ch Did you have a chance to look at this?

**Comment by pitr-ch** [comment]
Hi sorry, I was otherwise occupied. I think using Promises might be a better option here. Multiple executors are not necessary. Here is an example:
```ruby
jobs.map { |j| Concurrent::Promise.execute(&j) }
# block until all are executed
Concurrent::Promise.all?(*jobs).wait!
jobs.map(&:value) # => array of job results
```
or using new promises implementation
```ruby
require 'concurrent-edge'
jobs.map { |j| Concurrent::Promises.future(&j) }
# block until all are executed
Concurrent::Promises.zip(*jobs).value! # => array of job results
```

**Comment by ankuriitk** [comment]
@pitr-ch : thanks for your reply. I think the problem is that the "nesting" is dynamic i.e. a promise internally may have another promise. So I can't chain them statically. The question is that if a promise "spawns" another promise, is deadlock possible if one uses a single global executor?

**Comment by ankuriitk** [comment]
@pitr-ch : FYI, my issue was https://github.com/ruby-concurrency/concurrent-ruby/issues/223. I had a proc that relied on variables defined in outer scope and I was using that proc directly in Future instead of passing the proc as an argument. Otherwise Future/Barrier combination is working fine.

**Comment by pitr-ch** [comment]
Deadlock is possible but improbable. There would have to be around 2000 (thread os limit) blocked promises execution simultaneously and all of the promises would have to be blocked by work which is thill in the queue of the thread-pool. Flat operation on old Promise block a thread though so it increases the danger. New Promises do not block a thread by doing a flat operation so the flat operation can be used safely without increased the danger of deadlock.

**Comment by pitr-ch** [comment]
Please reopen if you have further questions.
